Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp

```diff
@@ -3067,6 +3071,91 @@ auto AirIRGenerator::addI31GetU(ExpressionType ref, ExpressionType& result) -> P
     return { };
 }
 
+auto AirIRGenerator::addStructNew(uint32_t typeIndex, Vector<ExpressionType>& args, ExpressionType& result) -> PartialResult
+{
+    ASSERT(typeIndex < m_info.typeCount());
+    result = tmpForType(Type { TypeKind::Ref, Nullable::No, m_info.typeSignatures[typeIndex]->index() });
+    auto jsInstance = g64();
+    append(Move, Arg::addr(instanceValue(), Instance::offsetOfOwner()), jsInstance);
+    // FIXME: inline the allocation.
+    // https://bugs.webkit.org/show_bug.cgi?id=244388
+    emitCCall(&JSWebAssemblyStruct::create, result, jsInstance, addConstant(Types::I32, typeIndex));
+
+    const auto& structType = *m_info.typeSignatures[typeIndex]->template as<StructType>();
+    for (unsigned i = 0; i < args.size(); ++i) {
+        auto status = addStructSet(result, structType, i, args[i]);
+        if (!status)
+            return status;
+    }
+
+    return { };
+}
+
+auto AirIRGenerator::addStructGet(ExpressionType structReference, const StructType& structType, uint32_t fieldIndex, ExpressionType& result) -> PartialResult
+{
+    uint32_t offset = fixupPointerPlusOffset(structReference, JSWebAssemblyStruct::getFieldOffset(fieldIndex, &structType));
+    Arg addrArg = Arg::addr(structReference, offset);
+
+    const auto& fieldType = structType.field(fieldIndex).type;
+    result = tmpForType(fieldType);
+    switch (fieldType.kind) {
+    case TypeKind::I32: {
+        appendEffectful(Move32, addrArg, result);
+        break;
+    }
+    case TypeKind::Funcref:
+    case TypeKind::Externref:
+    case TypeKind::Ref:
+    case TypeKind::RefNull:
+    case TypeKind::I64:
+        appendEffectful(Move, addrArg, result);
+        break;
+    case TypeKind::F32:
+        appendEffectful(MoveFloat, addrArg, result);
+        break;
+    case TypeKind::F64:
+        appendEffectful(MoveDouble, addrArg, result);
+        break;
+    default:
+        RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    return { };
+}
+
+auto AirIRGenerator::addStructSet(ExpressionType structReference, const StructType& structType, uint32_t fieldIndex, ExpressionType value) -> PartialResult
+{
+    uint32_t offset = fixupPointerPlusOffset(structReference, JSWebAssemblyStruct::getFieldOffset(fieldIndex, &structType));
+    Arg addrArg = Arg::addr(structReference, offset);
+
+    const auto& fieldType = structType.field(fieldIndex).type;
+    switch (fieldType.kind) {
+    case TypeKind::I32:
+        append(Move32, value, addrArg);
+        break;
+    case TypeKind::Funcref:
+    case TypeKind::Externref:
+    case TypeKind::Ref:
+    case TypeKind::RefNull:
+    case TypeKind::I64: {
+        if (isRefType(fieldType))
+            emitWriteBarrierForJSWrapper();
```

Comment:
This colors the whole Instance, not just the target struct cell, right? We only need to color the target struct cell here,

Reply:
Right, but we mark a whole instance for global.set too. Weird, maybe there is a reason to do this, but I don't know it.

Reply:
Meanwhile I fixed struct.set implementation to use a more precise write barrier, for Air and B3.